fog/ledger/server/src/db_fetcher.rs

```diff
@@ -164,62 +163,51 @@ impl<
     pub fn run(mut self) {
         log::info!(self.logger, "Db fetcher thread started.");
         let block_range = self.sharding_strategy.get_block_range();
-        let mut next_block_index = 0;
+        let mut next_block_index = block_range.start_block;
         loop {
-            if self.stop_requested.load(Ordering::SeqCst) {
-                log::info!(self.logger, "Db fetcher thread stop requested.");
-                break;
+            loop {
+                let num_blocks = self.load_block_data(&mut next_block_index);
+
+                let end = min(num_blocks, block_range.end_block);
+
+                if next_block_index < end.saturating_sub(BLOCKS_BEHIND) {
+                    self.readiness_indicator.set_unready();
+                } else {
+                    self.readiness_indicator.set_ready();
+                }
+
+                if end <= next_block_index {
+                    break;
+                }
+
+                if self.stop_requested.load(Ordering::SeqCst) {
+                    break;
+                }
             }
 
-            // Each call to load_block_data attempts to load one block for each known
-            // invocation. We want to keep loading blocks as long as we have data to load,
-            // but that could take some time which is why the loop is also gated
-            // on the stop trigger in case a stop is requested during loading.
-            while self.load_block_data(&mut next_block_index, &block_range)
-                && !self.stop_requested.load(Ordering::SeqCst)
-            {
-                // Hack: If we notice that we are way behind the ledger, set ourselves unready
-                match self.block_provider.num_blocks() {
-                    Ok(num_blocks) => {
-                        // if there are > BLOCKS_BEHIND *available* blocks we haven't loaded yet,
-                        // set unready
-                        if min(num_blocks, block_range.end_block) > next_block_index + BLOCKS_BEHIND
-                        {
-                            self.readiness_indicator.set_unready();
-                        }
-                    }
-                    Err(err) => {
-                        log::error!(self.logger, "Could not get num blocks from db: {}", err);
-                    }
-                };
+            if !block_range.contains(next_block_index) {
```

Comment:
Any reason this can't be `if next_block_index == block_range.end_block {`?

Reply:
it could be if desired. 
I generally lean towards bounds checking versus in/equality 
I originally had this as `if block_range.end_block <= next_block_index`, but thought the "contains" might be clearer intent, however that "!" in front can be missed...

Reply:
similar to above, I'm not opposed to discussing further, but "good enough for now" to get a fast turn around on the release